docs/source/guide/resources.rst

```diff
@@ -199,22 +199,32 @@ keyword arguments. Examples of waiters include::
 
 
 Multithreading and multiprocessing
---------------------------------
-It is recommended to create a resource instance for each thread / process in a multithreaded or multiprocess application rather than sharing a single instance among the threads / processes. For example::
+----------------------------------
+SDK object instances are **not** thread safe in almost all cases and should not be shared across threads or processes. You should create a new SDK object for each thread or process::
 
     import boto3
     import boto3.session
     import threading
 
     class MyTask(threading.Thread):
         def run(self):
+            # Here we create a new session per thread
             session = boto3.session.Session()
+
+            # Next, we create a resource client using our thread's session object
             s3 = session.resource('s3')
-            # ... do some work with S3 ...
+
+            # Put your thread-safe code here
 
 In the example above, each thread would have its own Boto3 session and its own instance of the S3 resource. This is a good idea because resources contain shared data when loaded and calling actions, accessing properties, or manually loading or reloading the resource can modify this data.
 
 .. note::
-    Resources are **not** thread safe. These special classes contain additional meta data that cannot be shared between threads. When using a Resource, it is recommended to instantiate a new Resource for each thread, as is shown in the example above. 
-    
-    Low-level clients **are** thread safe. When using a low-level client, it is recommended to instantiate your client then pass that client object to each of your threads. 
+    Resources are **not** thread safe. These special classes contain additional meta data that cannot be shared between threads. When using a Resource, it is recommended to instantiate a new Resource for each thread, as is shown in the example above.
+
+    Sessions are **not** thread safe. This is due to shared meta data, similar to Resource.
+
+    Low-level clients **are** thread safe in a **threaded environment** in all cases when using the client's attached operation methods (i.e. ``s3.get_object``). However, clients are **not** thread safe if you are explicitly using the client to interact with the underlying Session (i.e. using ``s3.meta``, ``s3.exceptions``, or ``s3.waiter_names``).
```

Comment:
For clients, we'll need to get a new section added to the bottom of the [guide](https://github.com/boto/boto3/blob/develop/docs/source/guide/clients.rst), similar to Resource and Session.

Feel free to riff on this if you feel there's better phrasing or layout, but something along these lines would be a great start:

## Multithreading or multiprocessing with Clients
Unlike Resources and Sessions, clients **are** generally _thread-safe_. There are some caveats, defined below, to be aware of though.

### Caveats

**Multi-Processing:** While clients are _thread-safe_, they cannot be shared across processes due to their networking implementation. Doing so may lead to incorrect response ordering when calling services.

**Shared Metadata:** Clients expose metadata to the end user through a few attributes (namely `meta`, `exceptions` and `waiter_names`). These are safe to read but any mutations should not be considered thread-safe.

**Custom [Botocore Events](https://botocore.amazonaws.com/v1/documentation/api/latest/topics/events.html):** Botocore (the library Boto3 is built on) allows advanced users to provide their own custom event hooks which may interact with boto3's client. The majority of users will not need to use these interfaces, but those that do should no longer consider their clients thread-safe without careful review.

### General Example

```python
import boto3.session
from concurrent.futures import ThreadPoolExecutor

def do_s3_task(client, task_definition):
    # Put your thread-safe code here

def my_workflow():
    # Create a session and use it to make our client
    session = boto3.session.Session()
    s3_client = session.client('s3')

    # Define some work to be done, this can be anything
    my_tasks = [ ... ]

    # Dispatch work tasks with our s3_client
    with ThreadPoolExecutor(max_workers=8) as executor:
        futures = [executor.submit(do_s3_task, s3_client, task) for task in my_tasks]
```